### ISSUE TITLE
Xpetra/Epetra: link error in panzer test for intel debug static

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#4038]
When I try to link packages/panzer/adapters-stk/test/solver/PanzerAdaptersSTK_solver.exe I get a missing tpetra symbol for intel debug static, but intel opt static works..  No idea why.  This is with a RHEL7 machine using all sems and Ross' new build scripts.  

Here is the error message I see, no idea why the xpetra epetra map is looking for something tpetra??  

```
packages/xpetra/src/libxpetra.a(Xpetra_EpetraMap.cpp.o): In function `Tpetra::Details::LocalMap<int, long, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >::LocalMap()':

```

Here is my link line

```
: && /projects/sems/install/rhel7-x86_64/sems/compiler/intel/17.0.1/openmpi/1.10.1/bin/mpicxx  --std=c++11 -g -fopenmp     -g   packages/panzer/adapters-stk/test/solver/CMakeFiles/PanzerAdaptersSTK_solver.dir/solver.cpp.o packages/panzer/adapters-stk/test/solver/CMakeFiles/PanzerAdaptersSTK_solver.dir/__/__/__/__/phalanx/test/Utilities/Phalanx_UnitTestMain.cpp.o  -o packages/panzer/adapters-stk/test/solver/PanzerAdaptersSTK_solver.exe  packages/panzer/adapters-stk/src/libpanzer-stk.a packages/muelu/adapters/libmuelu-adapters.a packages/muelu/src/Interface/libmuelu-interface.a packages/muelu/src/libmuelu.a packages/teko/src/libteko.a packages/ifpack2/adapters/libifpack2-adapters.a packages/ifpack2/src/libifpack2.a packages/seacas/libraries/ioss/src/main/libio_info_lib.a packages/seacas/libraries/ioss/src/init/libIonit.a packages/seacas/libraries/ioss/src/transform/libIotr.a packages/seacas/libraries/ioss/src/heartbeat/libIohb.a packages/seacas/libraries/ioss/src/gen_struc/libIogs.a packages/seacas/libraries/ioss/src/generated/libIogn.a packages/seacas/libraries/ioss/src/visualization/libIovs.a packages/seacas/libraries/ioss/src/pamgen/libIopg.a packages/seacas/libraries/ioss/src/exo_fac/libIoexo_fac.a packages/seacas/libraries/ioss/src/exo_fpp/libIofx.a packages/seacas/libraries/ioss/src/exodus/libIoex.a packages/seacas/libraries/ioss/src/libIoss.a packages/seacas/libraries/exodus/libexodus.a packages/panzer/disc-fe/src/libpanzer-disc-fe.a packages/panzer/dof-mgr/src/libpanzer-dof-mgr.a packages/phalanx/src/libphalanx.a packages/panzer/core/src/libpanzer-core.a packages/piro/src/libpiro.a packages/muelu/adapters/libmuelu-adapters.a packages/muelu/src/Interface/libmuelu-interface.a packages/muelu/src/libmuelu.a packages/intrepid2/src/libintrepid2.a packages/sacado/src/libsacado.a packages/tempus/src/libtempus.a packages/rythmos/src/librythmos.a packages/nox/src-loca/src-thyra/liblocathyra.a packages/nox/src-loca/src-epetra/liblocaepetra.a packages/nox/src-loca/src-lapack/liblocalapack.a packages/nox/src-loca/src/libloca.a packages/nox/src-epetra/libnoxepetra.a packages/nox/src-lapack/libnoxlapack.a packages/nox/src/libnox.a packages/teko/src/libteko.a packages/ifpack2/adapters/libifpack2-adapters.a packages/ifpack2/src/libifpack2.a packages/zoltan2/src/libzoltan2.a packages/anasazi/tpetra/src/libanasazitpetra.a packages/anasazi/epetra/util/ModeLaplace/libModeLaplace.a packages/anasazi/epetra/src/libanasaziepetra.a packages/anasazi/src/libanasazi.a packages/stk/stk_io/stk_io/util/libstk_io_util.a packages/stk/stk_io/stk_io/libstk_io.a packages/seacas/libraries/ioss/src/main/libio_info_lib.a packages/seacas/libraries/ioss/src/init/libIonit.a packages/seacas/libraries/ioss/src/transform/libIotr.a packages/seacas/libraries/ioss/src/heartbeat/libIohb.a packages/seacas/libraries/ioss/src/gen_struc/libIogs.a packages/seacas/libraries/ioss/src/generated/libIogn.a packages/seacas/libraries/ioss/src/visualization/libIovs.a packages/seacas/libraries/ioss/src/pamgen/libIopg.a packages/seacas/libraries/ioss/src/exo_fac/libIoexo_fac.a packages/seacas/libraries/ioss/src/exo_fpp/libIofx.a packages/seacas/libraries/ioss/src/exodus/libIoex.a packages/seacas/libraries/ioss/src/libIoss.a packages/pamgen/src/libpamgen_extras.a packages/pamgen/src/libpamgen.a packages/stk/stk_mesh/stk_mesh/base/libstk_mesh_base.a packages/shards/src/libshards.a packages/stk/stk_topology/stk_topology/libstk_topology.a packages/stk/stk_util/stk_util/use_cases/libstk_util_use_cases.a packages/stk/stk_util/stk_util/registry/libstk_util_registry.a packages/stk/stk_util/stk_util/diag/libstk_util_diag.a packages/stk/stk_util/stk_util/environment/libstk_util_env.a packages/stk/stk_util/stk_util/parallel/libstk_util_parallel.a packages/stk/stk_util/stk_util/util/libstk_util_util.a /projects/sems/install/rhel7-x86_64/sems/tpl/boost/1.59.0/intel/17.0.1/base/lib/libboost_program_options.so /projects/sems/install/rhel7-x86_64/sems/tpl/boost/1.59.0/intel/17.0.1/base/lib/libboost_system.so packages/seacas/libraries/aprepro_lib/libaprepro_lib.a packages/seacas/libraries/exodus/libexodus.a -L/projects/sems/install/rhel7-x86_64/sems/tpl/hdf5/1.8.12/intel/17.0.1/openmpi/1.10.1/parallel/lib -L/projects/sems/install/rhel7-x86_64/sems/tpl/boost/1.59.0/intel/17.0.1/base/lib -L/projects/sems/install/rhel7-x86_64/sems/tpl/netcdf/4.4.1/intel/17.0.1/openmpi/1.10.1/exo_parallel/lib -L/lib -lboost_program_options -lboost_system -lnetcdf -lpnetcdf -lhdf5 -lcurl -lhdf5_hl -lz packages/stratimikos/src/libstratimikos.a packages/stratimikos/adapters/belos/src/libstratimikosbelos.a packages/stratimikos/adapters/amesos2/src/libstratimikosamesos2.a packages/stratimikos/adapters/aztecoo/src/libstratimikosaztecoo.a packages/stratimikos/adapters/amesos/src/libstratimikosamesos.a packages/stratimikos/adapters/ml/src/libstratimikosml.a packages/stratimikos/adapters/ifpack/src/libstratimikosifpack.a packages/amesos2/src/libamesos2.a packages/ml/src/libml.a packages/galeri/src-xpetra/libgaleri-xpetra.a packages/galeri/src-epetra/libgaleri-epetra.a packages/ifpack/src/libifpack.a packages/amesos/src/libamesos.a packages/common/auxiliarySoftware/SuiteSparse/src/libtrilinosss.a packages/belos/xpetra/src/libbelosxpetra.a packages/belos/tpetra/src/libbelostpetra.a packages/belos/epetra/src/libbelosepetra.a packages/belos/src/libbelos.a packages/xpetra/sup/libxpetra-sup.a packages/xpetra/src/libxpetra.a packages/thyra/adapters/tpetra/src/libthyratpetra.a packages/thyra/adapters/epetraext/src/libthyraepetraext.a packages/epetraext/src/libepetraext.a packages/thyra/adapters/epetra/src/libthyraepetra.a packages/thyra/core/src/libthyracore.a packages/rtop/src/librtop.a packages/tpetra/core/ext/libtpetraext.a packages/tpetra/core/inout/libtpetrainout.a packages/tpetra/core/src/libtpetra.a packages/kokkos-kernels/src/libkokkoskernels.a packages/kokkos/algorithms/src/libkokkosalgorithms.a packages/kokkos/containers/src/libkokkoscontainers.a packages/tpetra/classic/LinAlg/libtpetraclassiclinalg.a packages/tpetra/classic/NodeAPI/libtpetraclassicnodeapi.a packages/tpetra/classic/src/libtpetraclassic.a packages/aztecoo/src/libaztecoo.a packages/triutils/src/libtriutils.a packages/epetra/src/libepetra.a packages/teuchos/kokkoscomm/src/libteuchoskokkoscomm.a packages/teuchos/kokkoscompat/src/libteuchoskokkoscompat.a packages/teuchos/remainder/src/libteuchosremainder.a packages/teuchos/numerics/src/libteuchosnumerics.a packages/teuchos/comm/src/libteuchoscomm.a packages/teuchos/parameterlist/src/libteuchosparameterlist.a packages/teuchos/parser/src/libteuchosparser.a -mkl packages/teuchos/core/src/libteuchoscore.a packages/kokkos/core/src/libkokkoscore.a -ldl packages/zoltan/src/libzoltan.a -lm && :

```

@trilinos/<teamName>

## Your Environment

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by rppawlo** [comment]
@jhux2  - I seem to recall something similar being reported in another issue recently. Do you who I should ping for this?

**Comment by jhux2** [comment]
@rppawlo That's me :).   Can I just follow the ATDM wiki instructions to reproduce this?

**Comment by rppawlo** [comment]
I believe so. It's the RHEL 7 build with intel compilers. You may have to add a flag to build static libraries though. Not sure if static is part of the issue.

**Comment by jhux2** [comment]
`Xpetra_EpetraMap.hpp` is internally using `Tpetra::Details::FixedHashTable`.  In the current build, GO=long is disabled.  `Xpetra_EpetraMap` has a specialization for GO=long.  I suspect that this is getting compiled, resulting in the missing symbol.  Next question -- why is the specialization getting compiled?

**Comment by jhux2** [comment]
`Xpetra_EpetraMap.hpp` derives from `Xpetra::Map`.  The latter has the virtual method `getLocalMap` whose return type is a `Tpetra::LocalMap` *if* Tpetra and Xpetra's KokkosRefactor branch are both enabled.  (See PR #1669 for some context.)

In the current build, `HAVE_TPETRA_INST_INT_LONG` is not defined.  However, because both Tpetra and Xpetra's KokkosRefactor branch are enabled, `Xpetra::EpetraMap` still depends on `Tpetra::LocalMap`, which in turn needs a `FixedHashTable`.

**Comment by mhoemmen** [comment]
@jhux2 I'm confused about the relevance of #1669.  There shouldn't ever be any Tpetra + `GO=long` happening if `GO=long` is not enabled in Tpetra.  Epetra64 only uses (the equivalent of) `GO = long long`.  Thus, how could it be that you need FixedHashTable for `GO=long`?

**Comment by jhux2** [comment]
@mhoemmen You're correct, Tpetra is not enabling `GO=long`.  However, there appears to be some downstream Xpetra code that requires it.

**Comment by mhoemmen** [comment]
@jhux2 wrote:
> However, there appears to be some downstream Xpetra code that requires it.

Is the plan, then, to fix said downstream Xpetra code?  Or does Xpetra really need FixedHashTable with keys of type `long`?

**Comment by bathmatt** [comment]
Any progress on this?  We've ran into a bug we need to fix using intel compiler that we can't figure out in opt code..

**Comment by jhux2** [comment]
@bathmatt I haven't looked at this in a while.  I'll pick it back up.

**Comment by bartlettroscoe** [comment]
Why are we not seeing this in the ATDM Trilinos builds?   Are we missing an intel debug build?

**Comment by bathmatt** [comment]
Any status update?

**Comment by bartlettroscoe** [comment]
@bathmatt, what is the exact reproducing build name?  is it just `intel-debug`?  Is this `openmp` or `serial`?

We should set up a nightly build to reproduce this so that Trilinos developers can reproduce and see if they fix things.

**Comment by bathmatt** [comment]
intel-debug-openmp-shared

**Comment by bathmatt** [comment]
But I'm 99% sure static does it too, at least it used to

**Comment by bartlettroscoe** [comment]
FYI: I removed the `ATDM ENV Issue` label since I am not sure this is an env issue that is a defect in the compiler per say.  (We generally only use  `ATDM ENV Issue` only for issues that will only be fixed by upgrading or fixing something in the env which I don't think is the case here).

I am standing up a nightly `sems-rhel7-intel-17.0.1-openmp-complex-shared-debug` build that will post to CDash.  Hopefully that will help raise the visibility of this. 

@bathmatt, would you say this is a blocker for EMPIRE?

**Comment by bartlettroscoe** [comment]
FYI: The PR #4675 added a new `Trilinos-atdm-sems-rhel7-intel-17.0.1-openmp-complex-shared-debug` build that shows a bunch of build errors as shown [here](https://testing.sandia.gov/cdash-dev-view/index.php?project=Trilinos&parentid=4772755).  We will create a new ATDM Trilinos GitHUb issue with the exact current build errors and reproducability instructions.   Yes that is a complex build but the link errors don't look to have anything to with complex as they show:

```
packages/xpetra/src/libxpetra.so.12.13: undefined reference to `Tpetra::Details::FixedHashTable<long, int, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace> >::FixedHashTable()'
```

**Comment by mhoemmen** [comment]
@bartlettroscoe See https://github.com/trilinos/Trilinos/issues/4038 for this Xpetra link error.

**Comment by bartlettroscoe** [comment]
@mhoemmen said:

> @bartlettroscoe See #4038 for this Xpetra link error.

This is #4038 :-)

**Comment by mhoemmen** [comment]
@bartlettroscoe my bad ;-)

**Comment by pwxy** [comment]
This bug screwed me yesterday on the CEE LAN (RHEL6 machine)

Currently I "source ${WORKSPACE}/Trilinos/cmake/std/atdm/load-env.sh rhel6_intel-debug-serial", but that omits "-DNDEBUG" and I get linking errors such as:

packages/xpetra/src/libxpetra.a(Xpetra_EpetraMap.cpp.o): In function `Tpetra::Details::LocalMap<int, long, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace> >::LocalMap()':
/scratch/ptlin/empire/trilinos/Trilinos20190303/build/rhel6_intel_debug_serial/../../packages/tpetra/core/src/Tpetra_Details_LocalMap.hpp:77: undefined reference to `Tpetra::Details::FixedHashTable<long, int, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace> >::FixedHashTable()'

When I build with target "rhel6_intel-opt-serial" which includes "-DNDEBUG" I don't get the above linking error, so I assume that the above link error is due to code within a DEBUG block.

**Comment by bartlettroscoe** [comment]
@pwxy, is this the build error you are seeing?

**Comment by pwxy** [comment]
@bartlettroscoe yes.  Turns out adding -DNDEBUG did not help the rhel6_intel-debug error.

**Comment by bartlettroscoe** [comment]
CC: @bathmatt, @tcfisher, @pwxy 

@srajama1 (Trilinos Linear Solver Product Lead),

It seems like the priority of this issue is getting elevated by the ATDM APPs, at least EMPIRE?

**Comment by srajama1** [comment]
@bartlettroscoe : Xpetra is under data services. @kddevin is a better contact than me. I will check with Xpetra team as @kddevin is currently out of office.

**Comment by jhux2** [comment]
@srajama1 The MueLu team will take this up either later today or tomorrow.

**Comment by srajama1** [comment]
@jhux2 : Thank you !

**Comment by jhux2** [comment]
See PR #4863.

**Comment by bartlettroscoe** [comment]
CC: @bathmatt, @micahahoward, @pwxy, @rppawlo, @sebrowne,, @tcfisher, @fryeguy52 

Just noticed that these build errors in the build `Trilinos-atdm-sems-rhel7-intel-17.0.1-openmp-complex-shared-debug` seem to be gone starting on 4/13/2019 as shown in [this query](https://testing.sandia.gov/cdash-dev-view/index.php?project=Trilinos&date=2019-04-22&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-sems-rhel7-intel-17.0.1-openmp-complex-shared-debug&field2=buildstarttime&compare2=83&value2=2019-04-01).

@pwxy, can you confirm that the EMPIRE intel debug build is working?

I will run a SPARC build against an updated version of Trilinos 'develop' (the SPARC Trilinos Integration builds are passing as of yesterday for the first time in a long time).

We will go ahead and start monitoring this build and get the test failures cleaned up (but that will be a separate story).

**Comment by bartlettroscoe** [comment]
Looks like the intel-17.0.1 debug build of SPARC complete on a CEE RHEL6 machine.  I just tested SPARC' master' against Trilinos 'develop' on my CEE machine 'ceerws1113' pulled a few hours ago by running:

```
$ cd sparc/
$ gitdist pull

$ cd Trilinos/
$ git checkout github-develop
$ git pull
$ cd ...

$ env  \
    SPARC_TRIL_BUILDS_LIST=cee-rhel6_intel-17.0.1_intelmpi-5.1.2_serial_static_dbg \
     SPARC_TRIL_SKIP_NATIVE_BUILD=1 \
  ./sparc-tril-dev-scripts/run_builds_and_tests.sh 

Running the builds:
  cee-rhel6_intel-17.0.1_intelmpi-5.1.2_serial_static_dbg

./sparc-tril-dev-scripts/build_and_test_atdm_trilinos_and_sparc.sh \
   cee-rhel6_intel-17.0.1_intelmpi-5.1.2_serial_static_dbg "" ""


real    199m1.131s
user    705m57.585s
sys     139m34.894s

    Trilinos_DIR=/scratch/rabartl/SPARC.base/sparc/Trilinos/cee-rhel6_intel-17.0.1_intelmpi-5.1.2_serial_static_dbg/lib/cmake/Trilinos

98% tests passed, 5 tests failed out of 280
```

I have never run a full debug build of SPARC before so I don't know if these 5 failing tests are related to Trilinos or not.

But at least the SPARC executable built and ran all of the tests.

**Comment by bartlettroscoe** [comment]
@pwxy, @bathmatt, @rppawlo,

Can someone on EMPIRE team confirm this is fixed?

**Comment by bathmatt** [comment]
its fixes
